src/accelerate/accelerator.py

```diff
@@ -1388,17 +1395,23 @@ def prepare_model(self, model: torch.nn.Module, device_placement: bool = None, e
 
         if self.native_amp:
             model._original_forward = model.forward
-            model_forward_func = model.forward.__func__ if hasattr(model.forward, "__func__") else model.forward
             autocast_context = get_mixed_precision_context_manager(self.native_amp, self.autocast_handler)
-            new_forward = autocast_context(model_forward_func)
-            if hasattr(model.forward, "__func__"):
-                model.forward = MethodType(new_forward, model)
-                model.forward = MethodType(convert_outputs_to_fp32(model.forward.__func__), model)
+            # NOTE: MS-AMP is special, and adds a `__func__` already to `model.forward`
```

Comment:
Good idea to add a comment here, but at least for me, it's not obvious what is happening here even with the comment.

Also, it seems to be that the code path for `self.fp8_backend == "MSAMP"` and the code path for `not hasattr(model.forward, "__func__")` are the same, is that right? If so, would it make sense to combine them into a single condition?

```python
if (self.fp8_backend == "MSAMP") or not hasattr(model.forward, "__func__")`:
    ...
else:
    ...
```